src/wechaty_puppet_hostie/puppet.py

```diff
@@ -155,9 +155,8 @@ def __init__(self, options: PuppetOptions, name: str = 'hostie_puppet'):
                 raise Exception('wechaty-puppet-hostie: token not found.')
             options.token = WECHATY_PUPPET_HOSTIE_TOKEN
 
-        if options.end_point is None \
-                and WECHATY_PUPPET_HOSTIE_ENDPOINT is not None:
-            options.end_point = WECHATY_PUPPET_HOSTIE_ENDPOINT
+        self.end_point: str = WECHATY_PUPPET_HOSTIE_ENDPOINT or ''
+        self.port: int = 0
```

Comment:
I think `end_point` and `port` shuold not be attributes of `puppet-hostie`. Because it's a type of configuration, so it should be in `options`. 

`Hostie Server` has supported dynamic port, so we can remove the default `port`. The changes below is my thought: 
```python
def _init_puppet(self):
    """
    start puppet channel contact_self_qr_code
    """
    log.info('init puppet')

    response = requests.get(
        f'https://api.chatie.io/v0/hosties/{self.options.token}'
    )

    if response.status_code != 200:
        raise Exception('hostie server is invalid ... ')

    data = response.json()
    if 'ip' not in data or data['ip'] == '0.0.0.0':
        raise Exception("can't find hostie server address")
    if 'port' not in data:
        raise Exception("can't find hostie server port")

    log.debug('get puppet ip address : <%s>', data)
    self.options.end_point = data['ip']
    log.info('init puppet hostie')

    self.channel = Channel(host=self.options.end_point, port=data['port'])
    self.puppet_stub = PuppetStub(self.channel)
```

How do you think about it ?

Reply:
But still if we have multiple bot running at a time, storing the endpoint and port in the global scope is still a problem.

Reply:
I think saving the option set end_point and port to the puppet object when init will be a better practise, I think modifying the option object is not a good idea, it should be readonly in this case.

Reply:
`endpoint` and `port` are stored in `hostie-puppet` instance, not in global scope. It's the specific instanace for every wechaty bot. 

Why we have multi bot? Because there are different `hostie-puppet`, which is connected with the specific token.

Reply:
For your consider about setting an end_point in the options object, I think the below treatment might be better, we load the end_point and default port into the puppet object when init.

```
    def _init_puppet(self):
        """
        start puppet channel contact_self_qr_code
        """
        log.info('init puppet')
        # load the end_point from the options object if set.
        if self.options.end_point or WECHATY_PUPPET_HOSTIE_ENDPOINT:
            self.end_point = self.options.end_point or WECHATY_PUPPET_HOSTIE_ENDPOINT
            self.port = 8788
        if not self.end_point:
            response = requests.get(
                f'https://api.chatie.io/v0/hosties/{self.options.token}'
            )

            if response.status_code != 200:
                raise Exception('hostie server is invalid ... ')

            data = response.json()
            if 'ip' not in data or data['ip'] == '0.0.0.0':
                raise Exception("can't find hostie server address")
            if 'port' in data:
                self.port = data['port']
            log.debug('get puppet ip address : <%s>', data)
            self.end_point = data['ip']
        log.info('init puppet hostie')

        self.channel = Channel(host=self.end_point, port=self.port)
        self.puppet_stub = PuppetStub(self.channel)
```